designs/aws-launch-template-network-interfaces.md

```diff
@@ -0,0 +1,129 @@
+    - AssociateCarrierIpAddress: false    # optional, Indicates whether to associate a Carrier IP address with eth0 for a new network interface.
+      AssociatePublicIpAddress: false     # optional, Indicates whether to assign a public IPv4 address to eth0 for a new network interface.
+      DeleteOnTermination: false          # optional, Indicates whether the network interface is deleted when the instance is terminated.
+      Description: "..."                  # optional, A description for the network interface.
+      DeviceIndex: 0                      # required, The device index for the network interface attachment.
+      InterfaceType: "interface"          # optional, The interface type for the network interface. To create an Elastic Fabric Adapter (EFA), specify efa
+      NetworkCardIndex: 0                 # optional, The index of the network card. Some instance types support multiple network cards. The primary network interface must be assigned to network card index 0. The default is network card index 0.
+      ipv4PrefixCount: 1                  # optional, The number of IPv4 prefixes that AWS automatically assigns to the network interface.
+      ipv6PrefixCount: 1                  # optional, The number of IPv6 prefixes that AWS automatically assigns to the network interface.
+    - { ... }
+```
+
+in addition, we will only support two networkInterfaces per node, this way we guarantee that the node will be able to
+run on all instance types.
+
+subnet and securityGroup will remain part of the `AWSNodeTemplate` and will be used for both network interfaces.
+in following iterations, as the `AWSNodeTemplate` move from `v1alpha1` to higher versions, we can add support for:
+
+```yaml
+subnetSelector: String
+securityGroupSelector: String
+```
+
+this way we can support different subnets and security groups for each network interface, we can also remove the limit
+of having only 2 network interfaces per LaunchTemplate.
+
+The [drift design doc #366](https://github.com/aws/karpenter-core/pull/366) suggests classifying CRD fields as  (1) Static, (2) Dynamic, or (3) Behavioral. in this
+context, the networkInterfaces would be classified as following:
+
+|                                                                | Static | Dynamic | Behavioral |
+|----------------------------------------------------------------|--------|---------|------------|
+| AWSNodeTemplate.NetworkInterfaces.length                       | x      |         |            |
+| AWSNodeTemplate.NetworkInterfaces.order                        | x      |         |            |
+| AWSNodeTemplate.NetworkInterfaces[@].AssociateCarrierIpAddress | x      |         |            |
+| AWSNodeTemplate.NetworkInterfaces[@].AssociatePublicIpAddress  | x      |         |            |
+| AWSNodeTemplate.NetworkInterfaces[@].DeleteOnTermination       |        |         | x          |
```

Comment:
I think this should be static as well. This is describable via an EC2 DescribeInstances call: https://docs.aws.amazon.com/cli/latest/reference/ec2/run-instances.html#:~:text=%22DeleteOnTermination%22%3A%20true%2C

Reply:
> I think this should be static as well. This is describable via an EC2 DescribeInstances call

True, I just thought it does not benefit much to have it. 
if the current instance has the value `false`, while the CRD changed to `true`. decommissioning the node for the drift will leave the networkInterface un deleted anyway. and vice versa. it makes more sense for an in-place drift fixing but this is not going to be supported.

Reply:
ah I see your point